src/cdk/testing/change-detection.ts

```diff
@@ -0,0 +1,112 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {BehaviorSubject, Subscription} from 'rxjs';
+
+/** Represents the status of auto change detection. */
+export interface AutoChangeDetectionStatus {
+  /** Whether auto change detection is disabled. */
+  isDisabled: boolean;
+  /**
+   * An optional callback, if present it indicates that change detection should be run immediately,
+   * while handling the status change. The callback should then be called as soon as change
+   * detection is done.
+   */
+  onDetectChangesNow?: () => void;
+}
+
+/** Subject used to dispatch and listen for changes to the auto change detection status . */
+const autoChangeDetectionSubject = new BehaviorSubject<AutoChangeDetectionStatus>({
+  isDisabled: false
+});
+
+/** The current subscription to `autoChangeDetectionSubject`. */
+let autoChangeDetectionSubscription: Subscription | null;
+
+/**
+ * The default handler for auto change detection status changes. This handler will be used if the
+ * specific environment does not install its own.
+ * @param status The new auto change detection status.
+ */
+function defaultAutoChangeDetectionHandler(status: AutoChangeDetectionStatus) {
+  status.onDetectChangesNow?.();
+}
+
+/**
+ * Allows a test `HarnessEnvironment` to install its own handler for auto change detection status
+ * changes.
+ * @param handler The handler for the auto change detection status.
+ */
+export function handleAutoChangeDetectionStatus(
+    handler: (status: AutoChangeDetectionStatus) => void) {
+  stopHandlingAutoChangeDetectionStatus();
+  autoChangeDetectionSubscription = autoChangeDetectionSubject.subscribe(handler);
+}
+
+/** Allows a `HarnessEnvironment` to stop handling auto change detection status changes. */
+export function stopHandlingAutoChangeDetectionStatus() {
+  autoChangeDetectionSubscription?.unsubscribe();
+  autoChangeDetectionSubscription = null;
+}
+
+/**
+ * Batches together triggering of change detection over the duration of the given function.
+ * @param fn The function to call with batched change detection.
+ * @param triggerBeforeAndAfter Optionally trigger change detection once before and after the batch
+ *   operation. If false, change detection will not be triggered.
+ * @return The result of the given function.
+ */
+async function batchChangeDetection<T>(fn: () => Promise<T>, triggerBeforeAndAfter: boolean) {
+  // If change detection batching is already in progress, just run the function.
+  if (autoChangeDetectionSubject.getValue().isDisabled) {
+    return await fn();
+  }
+
+  // If nothing is handling change detection batching, install the default handler.
+  if (!autoChangeDetectionSubscription) {
+    autoChangeDetectionSubject.subscribe(defaultAutoChangeDetectionHandler);
+  }
+
+  if (triggerBeforeAndAfter) {
+    await new Promise(resolve => autoChangeDetectionSubject.next({
+      isDisabled: true,
+      onDetectChangesNow: resolve,
+    }));
+    const result = await fn();
```

Comment:
If the specified function fails here, we will never restore the disabled state of the auto change-detection, and change detection will never run at all. This causes the CI failures you're seeing in the gulp jobs as those run all tests (like a usual CLI project) in a single Karma process. This is actually great as it unveiled this issue that could surface for consumers.

The fix is to wrap this with a `try/finally` and always restore the state of change detection. In tests, it can commonly happen that errors are expected & **silently handled**. In such situations consecutive tests just fail randomly with all kinds of errors as change detection for harnesses is disabled permanently (as seen with the CI failures for this PR; it was hard figuring out there is actually a "silent" error). It was a test in the radio harnesses using `expectAsync` and `toBeRejectedWith` 😄 

```ts
// If the function fails, we still want to restore the auto change detection state as
// otherwise change detection could be disabled unexpectedly forever. In some cases,
// errors are expected in tests and therefore silently captured. This would mean that
// consecutive tests relying on auto change detection could break without any hint about
// disabled auto change detection.
try {
  result = await fn();
} finally {
  await new Promise(resolve => autoChangeDetectionSubject.next({
    isDisabled: false,
    onDetectChangesNow: resolve,
  }));
}
```

We also need to do the same below with `triggerBeforeAndAfter = false`.

Reply:
For later reference: The select tests for example failed as `selectHarness.open` was not reflected due to disabled change detection. The options were undefined then and `.click` resulted in a runtime exception.

Reply:
Thank you so much for looking into this! That makes total sense. I was going crazy trying to debug this on Friday